008-auth.md

```diff
@@ -0,0 +1,254 @@
+    roles
+- greater authorization flexibility
+  - allow decoupling roles from key sets
+  - allow dynamic and arbitrary mappings of authentication data to roles
+
+Related issues:
+ - https://github.com/talos-systems/talos/issues/3306
+
+## Case Studies
+
+### Any larger organisation
+
+In general, passing authentication credentials around on the network is an
+anti-pattern.
+Once the set of managers of a system grows beyond one, greater flexibility in
+the distribution and management of key and assignment material is necessary.
+
+### Organisations with an existing SSO system
+
+When an organisation already has in place a common authentication system,
+we should be able to externalise the sourcing of our authentication to their
+system.
+However, we still need to be able to control mappings of those identities to
+Talos roles.
+
+## Motivation
+
+
+Why are we doing this? What use cases does it support? What is the expected
+outcome?
+
+# Assumptions
+
+What assumptions do we make about the elements this RFC touches?
+
+If this is a feature with significant memory overhead, do we assume end-users
+can all meet it?
+
+If this RFC is to make a security improvement, what threat model is it
+operating under? List any assumptions we make about the capabilities we feel
+we can reasonably make about potential adversaries.
+
+# Requirements
+
+What exactly is the outcome we want and want to enforce moving forward?
+
+List some MUST, MUST NOT, and SHOULD statements here that people will be
+expected to use as a quick reference to consider when doing engineering work.
+
+# Design
+
+This should be a simple bulleted explanation of the core process or outcome
+that we expect when this RFC is implemented.
+
+This could be a developer workflow, an automation workflow, a specific sequence
+of events, etc.
+
+Include any diagrams as appropriate, favoring text-based sources others can
+easily edit like mermaid, UML, etc. where possible, with external files sharing
+the same filename prefix as this document.
+
+# Examples
+
+Detail one or more hypothetical but plausible real-world scenarios where this
+RFC would offer a tangible benefit.
+
+# Drawbacks
+
+Provide a few bullets honestly detailing the downsides of implementing this.
+
+Will it slow down development? Cost more resources? Etc.
+
+# Alternatives
+
+List any alternatives others have tried, including the current solution if
+there is one, to help others quickly catch up on how we got here.
+
+# Scope
+
+List a few items that are clearly in scope, anything that we want to do that is
+explicitly deferred to future work.
+
+## Questions
+
+What open questions still exist that we can't answer right now?
+
+
+
+
+
+
+# Client authentication
+
+
+## Phase 1
+
+First, we need to break apart the `talosconfig` from the authentication PKI.
+This allows us to, among other things: 
+ - encourage the distribution of `talosconfig` without secret material included
```

Comment:
Probably out of scope for this RFC, but still worth to mention:

This also applies to machine node configs, which also contain CA keys, essentially the whole kingdom to the castle.

I tried to have terraform create with PKI configs, never realize the machine config on disk, directly passing along the derived config to $cloudProvider. However, this is right now pretty hard: https://github.com/siderolabs/talos/issues/5056#issuecomment-1054528830.

Ideally there should be a more automatic way to do this, other than manually editing config files with secrets.

Reply:
Yes.  In this change, we are wanting to break authentication from authorization.  We are also separately working toward separation of cluster PKI from machineconfig.
So it's out of scope for this change, but these changes will facilitate that, to an extent.